AdventOfCode.Library/Day13/Part1.fs

```diff
@@ -0,0 +1,99 @@
+module AdventOfCode.Library.Day13.Part1
+
+module Parser =
+    type Parser<'T> = char list -> ('T * char list) option
+
+    let map (f: 'T -> 'U) (p: 'T Parser) : 'U Parser =
+        fun s1 -> p s1 |> Option.map (fun (t, s2) -> f t, s2)
+
+    let succeed (t: 'T) : 'T Parser = fun s -> Some(t, s)
+
+    let lift (f: 'T -> 'U -> 'V) (p: 'T Parser) (q: 'U Parser Lazy) : 'V Parser =
+        fun s1 ->
+            match p s1 with
+            | None -> None
+            | Some(t, s2) ->
+                match q.Force () s2 with
+                | None -> None
+                | Some(u, s3) -> Some(f t u, s3)
+
+    let choose (p: 'T Parser) (u: 'T Parser Lazy) : 'T Parser =
```

Comment:
Laziness was also necessary here, otherwise the second parser would always be evaluated even if the first one succeeded.